pkg/controller/packetsampling/controller.go

```diff
@@ -0,0 +1,363 @@
+	packetSamplingInformer     crdinformers.PacketSamplingInformer
+	packetSamplingLister       crdlisters.PacketSamplingLister
+	packetSamplingListerSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	runningPacketSamplingsMutex sync.Mutex
+	runningPacketSamplings      map[uint8]string
+}
+
+func NewPacketSamplingController(client versioned.Interface, podInformer coreinformers.PodInformer, packetSamplingInformer crdinformers.PacketSamplingInformer) *Controller {
+	c := &Controller{
+		client:                     client,
+		podInformer:                podInformer,
+		podLister:                  podInformer.Lister(),
+		packetSamplingInformer:     packetSamplingInformer,
+		packetSamplingLister:       packetSamplingInformer.Lister(),
+		packetSamplingListerSynced: packetSamplingInformer.Informer().HasSynced,
+		queue:                      workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "packetsampling"),
+		runningPacketSamplings:     make(map[uint8]string),
+	}
+
+	// add handlers
+	packetSamplingInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addPacketSampling,
+			UpdateFunc: c.updatePacketSampling,
+			DeleteFunc: c.deletePacketSampling,
+		}, resyncPeriod,
+	)
+	return c
+}
+
+func (c *Controller) addPacketSampling(obj interface{}) {
+	ps := obj.(*crdv1alpha1.PacketSampling)
+	klog.V(2).Infof("Adding PacketSampling %s", ps.Name)
+	c.enqueuePacketSampling(ps)
+}
+
+func (c *Controller) updatePacketSampling(_, cur interface{}) {
+	ps := cur.(*crdv1alpha1.PacketSampling)
+	klog.V(2).Infof("Updating PacketSampling %s", ps.Name)
+	c.enqueuePacketSampling(ps)
+}
+
+func (c *Controller) deletePacketSampling(obj interface{}) {
+	ps := obj.(*crdv1alpha1.PacketSampling)
+	klog.V(2).Infof("Deleting PacketSampling %s", ps.Name)
+	c.deallocateTagForPS(ps)
+}
+
+func (c *Controller) enqueuePacketSampling(ps *crdv1alpha1.PacketSampling) {
+	c.queue.Add(ps.Name)
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.packetSamplingListerSynced) {
+		return
+	}
+
+	pss, err := c.packetSamplingLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list all PacketSamplings: %v", err)
+	}
+
+	for _, ps := range pss {
+		if ps.Status.Phase == crdv1alpha1.PacketSamplingRunning {
+			if err := c.occupyTag(ps); err != nil {
+				klog.Errorf("load packetsampling data plane tag failed: %+v+, %v", ps, err)
+			}
+		}
+	}
+
+	go func() {
+		wait.Until(c.checkPacketSamplingTimeout, timeoutCheckInterval, stopCh)
+	}()
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+
+}
+
+func (c *Controller) checkPacketSamplingTimeout() {
```

Comment:
You have two methods named `checkPacketSamplingTimeout`. I feel this one can be renamed to `requeuePacketSamplingsForTimeoutCheck`

Reply:
oudated.